Remove hard_limit_for_bookkeeping

PeterSolMS opened the pull request:
As it turns out, it's not so easy to set aside a hard limit for the GC's bookkeeping, because some regions will be partially comitted and so will need more bookkeeping information as a percentage.

So this removes the hard_limit_for_bookkeeping and associated fields, and instead commits memory for the mark array eagerly, even if BGC is not in progress.

PeterSolMS committed:
As it turns out, it's not so easy to set aside a hard limit for the GC's bookkeeping, because some regions will be partially comitted and so will need more bookkeeping information as a percentage.

So this removes the hard_limit_for_bookkeeping and associated fields, and instead commits memory for the mark array eagerly, even if BGC is not in progress.

cshung commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -4132,10 +4126,6 @@ class gc_heap
 
-    // This is what GC uses for its own bookkeeping.
-    PER_HEAP_ISOLATED
-    size_t current_total_committed_bookkeeping;
```
This field was there before I introduced `heap_hard_limit_for_bookkeeping`.

PeterSolMS replied:
Oops, ok, I'll put it back...

cshung commented on src/coreclr/gc/gc.cpp:
```diff
@@ -20340,20 +20285,17 @@ bool gc_heap::try_get_new_free_region()
+        dprintf (GC_TABLE_LOG, ("new seg %Ix, mark_array is %Ix",
+            heap_segment_mem (region), mark_array));
+        if (((region->flags & heap_segment_flags_ma_committed) == 0) &&
+            !commit_mark_array_new_seg (__this, region))
```
Shall we guard this under `gc_can_use_concurrent`?

Maoni0 replied:
`commit_mark_array_new_seg` shouldn't do anything if concurrent is disabled; although I'm looking at the code, we are not initializing `background_saved_highest_address` in `init_gc_heap` which we should. @PeterSolMS could you please add this (ie, init `background_saved_highest_address` and `background_saved_lowest_address` to 0 in `init_gc_heap` like we do with WKS GC)? we could still guard this with `gc_can_use_concurrent` just to indicate our intention but it's not essential.

PeterSolMS replied:
I just pushed changes with the initialization of `background_saved_lowest_address` and `background_saved_highest_address`, and the restoration of  `current_total_committed_bookkeeping`.

cshung reviewed (commented):
I liked this idea more than trying to set aside memory for bookkeeping. The one thing I am not sure about is that now we are committing the mark array for every region. If background GC is happening anyway, it might be fine because we will be committing them anyway, but what if we don't?

Maoni0 commented:
> If background GC is happening anyway, it might be fine because we will be committing them anyway, but what if we don't?

that _is_ the drawback with doing this. but if concurrent GC is enabled, it's very likely we'll need to do a BGC at some point so it's not too bad.

PeterSolMS committed:
Put current_total_committed_bookkeeping and its uses in dprintf statements back.

Initialize background_saved_lowest_address and background_saved_lowest_address.

Removed unrelated bug fix which is addressed by a separate PR.

PeterSolMS commented:
I think I found a hole in our logic - `clear_region_info` (called by `return_free_region`) will *decommit* the mark array in high memory load. When `get_free_region` tries to commit again, we'll fail and get the same AV as before. Perhaps this means we'll have to keep the mark array committed, i.e. disable the logic to decommit the mark array in high memory load?

PeterSolMS committed:
Do *not* decommit the mark array in high memory load in clear region info.

The issue is that GC needs at least one free region, and it needs the mark array committed. So when it tries to get a region at the end, and we cannot commit the mark array, we cannot actually get a free region. Currently we AV in this case, but there really isn't a good option to recover at this point. So it's better to keep the mark array committed and perhaps fail with an OOM exception.

PeterSolMS commented:
With this change, we get much closer to actually exhausting the available memory in hard limit cases than before.

In a test case with hard limit set to 0x13c1d000 = 471,859,200, we only get to 331,468,800 committed bytes before running out of memory. The reason is that the memory set aside for bookkeeping is not sufficient, so we run into a situation where the mark array for a region can not be committed.

With this change, we get to 471,842,816 committed bytes before running out. "!dumpheap -stat" reports 462,803,767 bytes in objects, of which 10,684,992 are in free objects, so that we have a total of 452,118,775 bytes in useful objects. 6,078,464 are used for bookkeeping purposes (mainly for the mark array).

PeterSolMS committed:
Merge with main.

PeterSolMS committed:
Merge branch 'main' into Remove_hard_limit_for_bookkeeping

frankbuckley commented:
> With this change, we get much closer to actually exhausting the available memory in hard limit cases than before.
> 
> In a test case with hard limit set to 0x13c1d000 = 471,859,200, we only get to 331,468,800 committed bytes before running out of memory...
> 
> With this change, we get to 471,842,816 committed bytes before running out...

Will this be backported to 7.0?
